src/dds_manager.cpp

```diff
@@ -1023,30 +1076,53 @@ bool DDSManager::replaceFilter(const std::string& topicName,
         targetTopic = topicGroup->topic;
     }
 
-    // Create the new data reader
+    // Create the new data reader, but first create a listener for it
     auto readerListener = std::make_unique<GenericReaderListener>();
     readerListener->SetHandler(m_rlHandler);
-    dataReader = topicGroup->subscriber->create_datareader(
-        targetTopic,
-        topicGroup->dataReaderQos,
-        readerListener.get(),
-        DDS::REQUESTED_INCOMPATIBLE_QOS_STATUS |
-        DDS::SUBSCRIPTION_MATCHED_STATUS |
-        DDS::SAMPLE_LOST_STATUS);
 
-    if (!dataReader)
+    if (readerListener)
     {
-        std::cerr << "Error creating data reader for '"
-            << topicName
-            << "'"
-            << std::endl;
+        // move replaces the original listener with the new listener and since there is no longer an existing
+        // reference to the orginial lister it deletes itself since it is a unique pointer.
+        // This must be done here rather than after the new reader, otherwise there will be duplicate reader
+        // listeners
+        topicGroup->m_readerListeners.emplace(readerName, std::move(readerListener));
+
+        //create the new data reader
+
+        dataReader = topicGroup->subscriber->create_datareader(
+            targetTopic,
+            topicGroup->dataReaderQos,
+            readerListener.get(),
```

Comment:
`readerListener` should not be used after the `std::move` on line 1089, except as the left-hand side of an assignment.

Instead this could directly reference the `topicGroup->m_readerListeners[readerName]` with that full expression or (probably better) with the iterator returned by `emplace()`

Reply:
Good find.  Using the  emplace result.